Fix scalebar behaviour at high/low zoom

brisvag opened the pull request:
# References and relevant issues
- closes https://github.com/napari/napari/issues/7956
- closes (kinda?) https://github.com/napari/napari/issues/6055

# Description

1. This issue has been bothering me for a long time and I finally figured it out. On main, if you have the scalebar enabled and zoom out beyond 1000, it will suddenly collapse to a single pixel.

Turns out, when the scalebar is dimensionless we need to ensure to keep the world pixel length to the *real* lenght, because we don't have unit multiples. Funny thing is, we already had this check in place, but for some reason only applied to high zooms, and not low zooms. Anyways, AFAICT this works.

2. The second issue is ultimately caused by our camera reporting a zoom of `1` when it reaches a low enough scale factor. This code points to https://github.com/napari/napari/issues/2875, but I can't figure out why.

EDIT: it seems to be https://github.com/napari/napari/pull/2958. I tried to remove this guard completely and I don't actually see that issue anymore... I think it might be solved, either by the fact that we moved a lot of camera updates to `on_draw` in https://github.com/napari/napari/pull/7870 specifically because the viewbox was degenerate, or for some other reason. Maybe we can just remove this?

brisvag committed:
fix scale bar at low zoom

brisvag committed:
stop reporting values as 1, report the lowest bound instead

TimMonko commented:
I'm getting a missing character instead of, I presume, a superscript `-`. It's fine in the other zoom direction 

<img width="217" height="95" alt="image" src="https://github.com/user-attachments/assets/b3391750-b139-4e1d-a573-c2bc63c02e6a" />

Otherwise seems to work great!

brisvag commented:
Dunno how to solve that :thinking: do we need to use a different vispy font? Can we even do that?

TimMonko committed:
Merge branch 'main' into fix/scale_bar

brisvag commented:
> How does it look if you add cells and zoom in super far? do you see the character correctly on Linux?

Same issue as you for me.

brisvag committed:
remove guard

brisvag commented:
> Also, I've removed the scale = 1 in at least a few places in the code during changes I've made and there have been no ill effects, so in other words, I do think we've moved beyond such bug

Yeah ok I'm convinced, let's remove.

psobolewskiPhD commented:
With this PR if I just swipe down on my trackpad to scroll, zooming in, i hit a traceback:
```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
File ~/Dev/miniforge3/envs/napari-dev/lib/python3.13/site-packages/vispy/app/backends/_qt.py:963, in CanvasBackendDesktop.paintGL(self=<vispy.app.backends._qt.CanvasBackendDesktop object>)
    961 # (0, 0, self.width(), self.height()))
    962 self._vispy_canvas.set_current()
--> 963 self._vispy_canvas.events.draw(region=None)
        self._vispy_canvas = <NapariSceneCanvas (PyQt6) at 0x13a01f0e0>
        self._vispy_canvas.events.draw = <vispy.util.event.EventEmitter object at 0x13f4247d0>
        self = <vispy.app.backends._qt.CanvasBackendDesktop object at 0x13f4258b0>
        self._vispy_canvas.events = <vispy.util.event.EmitterGroup object at 0x13f41ca50>    965 # Clear the alpha channel with QOpenGLWidget (Qt >= 5.4), otherwise the
    966 # window is translucent behind non-opaque objects.
    967 # Reference:  MRtrix3/mrtrix3#266
    968 if QT5_NEW_API or PYSIDE6_API or PYQT6_API:

File ~/Dev/miniforge3/envs/napari-dev/lib/python3.13/site-packages/vispy/util/event.py:453, in EventEmitter.__call__(self=<vispy.util.event.EventEmitter object>, *args=(), **kwargs={'region': None})
    450 if self._emitting > 1:
    451     raise RuntimeError('EventEmitter loop detected!')
--> 453 self._invoke_callback(cb, event)
        event = <DrawEvent blocked=False handled=False native=None region=None source=None sources=[] type=draw>
        self = <vispy.util.event.EventEmitter object at 0x13f4247d0>
        cb = <bound method VispyCanvas.on_draw of <napari._vispy.canvas.VispyCanvas object at 0x13a01dfd0>>    454 if event.blocked:
    455     break

File ~/Dev/miniforge3/envs/napari-dev/lib/python3.13/site-packages/vispy/util/event.py:471, in EventEmitter._invoke_callback(self=<vispy.util.event.EventEmitter object>, cb=<bound method VispyCanvas.on_draw of <napari._vispy.canvas.VispyCanvas object>>, event=<DrawEvent blocked=False handled=False native=None region=None source=None sources=[] type=draw>)
    469     cb(event)
    470 except Exception:
--> 471     _handle_exception(self.ignore_callback_errors,
        self = <vispy.util.event.EventEmitter object at 0x13f4247d0>
        cb = <bound method VispyCanvas.on_draw of <napari._vispy.canvas.VispyCanvas object at 0x13a01dfd0>>
        event = <DrawEvent blocked=False handled=False native=None region=None source=None sources=[] type=draw>
        (cb, event) = (<bound method VispyCanvas.on_draw of <napari._vispy.canvas.VispyCanvas object at 0x13a01dfd0>>, <DrawEvent blocked=False handled=False native=None region=None source=None sources=[] type=draw>)    472                       self.print_callback_errors,
    473                       self, cb_event=(cb, event))

File ~/Dev/miniforge3/envs/napari-dev/lib/python3.13/site-packages/vispy/util/event.py:469, in EventEmitter._invoke_callback(self=<vispy.util.event.EventEmitter object>, cb=<bound method VispyCanvas.on_draw of <napari._vispy.canvas.VispyCanvas object>>, event=<DrawEvent blocked=False handled=False native=None region=None source=None sources=[] type=draw>)
    467 def _invoke_callback(self, cb, event):
    468     try:
--> 469         cb(event)
        cb = <bound method VispyCanvas.on_draw of <napari._vispy.canvas.VispyCanvas object at 0x13a01dfd0>>
        event = <DrawEvent blocked=False handled=False native=None region=None source=None sources=[] type=draw>    470     except Exception:
    471         _handle_exception(self.ignore_callback_errors,
    472                           self.print_callback_errors,
    473                           self, cb_event=(cb, event))

File ~/Dev/napari/src/napari/_vispy/canvas.py:640, in VispyCanvas.on_draw(self=<napari._vispy.canvas.VispyCanvas object>, event=<DrawEvent blocked=False handled=False native=None region=None source=None sources=[] type=draw>)
    638 # sync all cameras
    639 for camera in (self.camera, *self.grid_cameras):
--> 640     camera.on_draw(event)
        camera = <napari._vispy.camera.VispyCamera object at 0x13f42a3c0>
        event = <DrawEvent blocked=False handled=False native=None region=None source=None sources=[] type=draw>    642 # The canvas corners in full world coordinates (i.e. across all layers).
    643 viewbox_corners_world = self._viewbox_corners_in_world

File ~/Dev/napari/src/napari/_vispy/camera.py:237, in VispyCamera.on_draw(self=<napari._vispy.camera.VispyCamera object>, _event=<DrawEvent blocked=False handled=False native=None region=None source=None sources=[] type=draw>)
    235 if not np.allclose(self.zoom, self._camera.zoom):
    236     with self._camera.events.zoom.blocker(self._on_zoom_change):
--> 237         self._camera.zoom = self.zoom
        self = <napari._vispy.camera.VispyCamera object at 0x13f42a3c0>
        self._camera.zoom = np.float64(inf)
        self._camera = Camera(center=(0.0, np.float64(3258.0383144293155), np.float64(2283.342099809898)), zoom=np.float64(inf), angles=(0.0, 0.0, 90.0), perspective=0.0, mouse_pan=True, mouse_zoom=True, orientation=(<DepthAxisOrientation.TOWARDS: 'towards'>, <VerticalAxisOrientation.DOWN: 'down'>, <HorizontalAxisOrientation.RIGHT: 'right'>))    238 if not np.allclose(self.perspective, self._camera.perspective):
    239     with self._camera.events.perspective.blocker(
    240         self._on_perspective_change
    241     ):

File ~/Dev/napari/src/napari/utils/events/evented_model.py:308, in EventedModel.__setattr__(self=Camera(center=(0.0, np.float64(3258.038314429315...n'>, <HorizontalAxisOrientation.RIGHT: 'right'>)), name='zoom', value=np.float64(inf))
    306     super().__setattr__(name, value)
    307     return
--> 308 with ComparisonDelayer(self):
        self = Camera(center=(0.0, np.float64(3258.0383144293155), np.float64(2283.342099809898)), zoom=np.float64(inf), angles=(0.0, 0.0, 90.0), perspective=0.0, mouse_pan=True, mouse_zoom=True, orientation=(<DepthAxisOrientation.TOWARDS: 'towards'>, <VerticalAxisOrientation.DOWN: 'down'>, <HorizontalAxisOrientation.RIGHT: 'right'>))    309     self._primary_changes.add(name)
    310     self._setattr_impl(name, value)

File ~/Dev/napari/src/napari/utils/events/evented_model.py:521, in ComparisonDelayer.__exit__(self=<napari.utils.events.evented_model.ComparisonDelayer object>, exc_type=None, exc_val=None, exc_tb=None)
    519 def __exit__(self, exc_type, exc_val, exc_tb):
    520     self._target._delay_check_semaphore -= 1
--> 521     self._target._check_if_values_changed_and_emit_if_needed()
        self._target = Camera(center=(0.0, np.float64(3258.0383144293155), np.float64(2283.342099809898)), zoom=np.float64(inf), angles=(0.0, 0.0, 90.0), perspective=0.0, mouse_pan=True, mouse_zoom=True, orientation=(<DepthAxisOrientation.TOWARDS: 'towards'>, <VerticalAxisOrientation.DOWN: 'down'>, <HorizontalAxisOrientation.RIGHT: 'right'>))
        self = <napari.utils.events.evented_model.ComparisonDelayer object at 0x1760dba70>
File ~/Dev/napari/src/napari/utils/events/evented_model.py:346, in EventedModel._check_if_values_changed_and_emit_if_needed(self=Camera(center=(0.0, np.float64(3258.038314429315...n'>, <HorizontalAxisOrientation.RIGHT: 'right'>)))
    343 with ComparisonDelayer(self):
    344     # Again delay comparison to avoid having events caused by callback functions
    345     for name, new_value in to_emit:
--> 346         getattr(self.events, name)(value=new_value)
        name = 'zoom'
        self = Camera(center=(0.0, np.float64(3258.0383144293155), np.float64(2283.342099809898)), zoom=np.float64(inf), angles=(0.0, 0.0, 90.0), perspective=0.0, mouse_pan=True, mouse_zoom=True, orientation=(<DepthAxisOrientation.TOWARDS: 'towards'>, <VerticalAxisOrientation.DOWN: 'down'>, <HorizontalAxisOrientation.RIGHT: 'right'>))
        new_value = np.float64(inf)
File ~/Dev/napari/src/napari/utils/events/event.py:752, in EventEmitter.__call__(self=<napari.utils.events.event.EventEmitter object>, *args=(), **kwargs={'value': np.float64(inf)})
    749     self._block_counter.update([cb])
    750     continue
--> 752 self._invoke_callback(cb, event if pass_event else None)
        event = <Event blocked=False handled=False native=None source=None sources=[] type='zoom'>
        self = <napari.utils.events.event.EventEmitter object at 0x139d9a780>
        cb = <bound method VispyScaleBarOverlay._on_zoom_change of <napari._vispy.overlays.scale_bar.VispyScaleBarOverlay object at 0x176026c10>>
        pass_event = False    753 if event.blocked:
    754     break

File ~/Dev/napari/src/napari/utils/events/event.py:790, in EventEmitter._invoke_callback(self=<napari.utils.events.event.EventEmitter object>, cb=<bound method VispyScaleBarOverlay._on_zoom_chan....overlays.scale_bar.VispyScaleBarOverlay object>>, event=None)
    788     self.disconnect(cb)
    789     return
--> 790 _handle_exception(
        self = <napari.utils.events.event.EventEmitter object at 0x139d9a780>
        event = None
        cb = <bound method VispyScaleBarOverlay._on_zoom_change of <napari._vispy.overlays.scale_bar.VispyScaleBarOverlay object at 0x176026c10>>
        (cb, event) = (<bound method VispyScaleBarOverlay._on_zoom_change of <napari._vispy.overlays.scale_bar.VispyScaleBarOverlay object at 0x176026c10>>, None)    791     self.ignore_callback_errors,
    792     self.print_callback_errors,
    793     self,
    794     cb_event=(cb, event),
    795 )

File ~/Dev/napari/src/napari/utils/events/event.py:779, in EventEmitter._invoke_callback(self=<napari.utils.events.event.EventEmitter object>, cb=<bound method VispyScaleBarOverlay._on_zoom_chan....overlays.scale_bar.VispyScaleBarOverlay object>>, event=None)
    777         cb(event)
    778     else:
--> 779         cb()
        cb = <bound method VispyScaleBarOverlay._on_zoom_change of <napari._vispy.overlays.scale_bar.VispyScaleBarOverlay object at 0x176026c10>>    780 except Exception as e:  # noqa: BLE001
    781     # dead Qt object with living python pointer. not importing Qt
    782     # here... but this error is consistent across backends
    783     if (
    784         isinstance(e, RuntimeError)
    785         and 'C++' in str(e)
    786         and str(e).endswith(('has been deleted', 'already deleted.'))
    787     ):

File ~/Dev/napari/src/napari/_vispy/overlays/scale_bar.py:125, in VispyScaleBarOverlay._on_zoom_change(self=<napari._vispy.overlays.scale_bar.VispyScaleBarOverlay object>, force=False)
    122     new_dim = self.overlay.length * self._unit.units
    123 else:
    124     # calculate the desired length as well as update the value and units
--> 125     target_world_pixels_rounded, new_dim = self._calculate_best_length(
        self = <napari._vispy.overlays.scale_bar.VispyScaleBarOverlay object at 0x176026c10>
        target_world_pixels = np.float64(0.0)    126         target_world_pixels
    127     )
    128     target_canvas_pixels = (
    129         target_world_pixels_rounded / scale_canvas2world
    130     )
    132 scale = target_canvas_pixels

File ~/Dev/napari/src/napari/_vispy/overlays/scale_bar.py:83, in VispyScaleBarOverlay._calculate_best_length(self=<napari._vispy.overlays.scale_bar.VispyScaleBarOverlay object>, desired_length=np.float64(0.0))
     75 factor = current_quantity / new_quantity
     77 # select value closest to one of our preferred values and also
     78 # validate if quantity is dimensionless and lower than 1 to prevent
     79 # the scale bar to extend beyond the canvas when zooming.
     80 # If the value falls in those conditions, we use the corresponding
     81 # preferred value but scaled to take into account the actual value
     82 # magnitude. See https://github.com/napari/napari/issues/5914
---> 83 magnitude_1000 = floor(log(new_quantity.magnitude, 1000))
        new_quantity = <Quantity(0.0, 'micrometer')>     84 scaled_magnitude = new_quantity.magnitude * 1000 ** (-magnitude_1000)
     85 index = bisect.bisect_left(PREFERRED_VALUES, scaled_magnitude)

ValueError: math domain error
```
On main I just get the behavior where the scale bar jumps back to a larger value, which is weird, but no tracebacks or errors. So I think the behavior with this PR is worse for the case of extreme zoom-in?

When I do the opposite, swipe to maximally zoom out, I hit:
`invalid value encountered in cast`
But I see no real ill effect. And in fact I get that on main too.

psobolewskiPhD commented:
So if I check out 20027401fcb717df788f574ce2b9d0e7912f8ece then the traceback I report above no longer happens.
So it seems like that guard is important in preventing the likely `log(0)` that is happening.

TimMonko commented:
> > Also, I've removed the scale = 1 in at least a few places in the code during changes I've made and there have been no ill effects, so in other words, I do think we've moved beyond such bug
> 
> Yeah ok I'm convinced, let's remove.

Is this what you ended up removing? The guard seems to be more precise than scale = 1, and things are broken now. I'm going to revert that commit and then hopefully @psobolewskiPhD can check it fixes things.

TimMonko committed:
Revert "remove guard"

This reverts commit a0a2b2906baca08cd9f597ebaa77e5f7b07582ff.

brisvag commented:
Funny, I didn't see this issue on my side. But yeah, we can leave it in.

TimMonko commented:
> Funny, I didn't see this issue on my side. But yeah, we can leave it in.

You have to go unbelievably zoomed in or zoomed out. Like far beyond practical limits, but you never know when someone just sends it with a smooth rolling mouse wheel (or Peter zooming on his trackpad)
